instant-distance/src/contiguous.rs

```diff
@@ -0,0 +1,240 @@
+use std::marker::PhantomData;
+
+use rand::{rngs::SmallRng, Rng, SeedableRng};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    types::{LayerId, Meta, INVALID},
+    Builder, Point, PointId,
+};
+
+pub trait PointDataSource: Sync {
+    fn decompose(&self) -> Vec<f32>;
```

Comment:
We definitely shouldn't be allocating for `decompose()`. What is the purpose of the `PointDataSource` trait and the `decompose()` method?

Reply:
I agree this would be better as a slice, but it's worth touching here to get a better idea for the intended design. 

`PointDataSource` is effectively the input type to our `Hnsw`, and is intended to be a way to convert what it's `impl`'d on to something that we can store, and later wrap in our "output" (or `Point` equivalent) type. In this iteration, that's `PointRef`. This means that the point data needs to be convertible (or "decomposed") to our internal storage (which for now is `Vec<f32>`).

Some open questions here:

- Prospectively, what does an API here look like that provides a generic input type that we can consume, and also provide access to, perhaps as a wrapper over slices of `f32` (or an `Element`, if that is workable). I'm also trying to keep the SIMD code workable.
- With the `Point` trait, we allowed `impl` of `fn distance` itself, but if we are looking to provide a general implementation internally (?), is that a concern that should remain externally-implementable? I could be misunderstanding the general goal for the API in this regard.

Reply:
At the Rust level, I think we can make `Point` pretty abstract: it provides a `distance() -> f32` method, an associated `Element` type and a `STRIDE` const. As I understand it, this should be sufficient on the Rust side. I suppose we can probably dump the `Clone` bound and do something where either `Point` exposes `as_slice(&self) -> &[Self::Element]` which we can then use to write in the `Hnsw::points` (or equivalent storage) or we can do it the other way around and add a `Point` method `store(&self, data: &mut [Self::Element])` method.

Reply:
Thanks, that was helpful.